rfcs/2022-08-22-11563-component-demo-mode.md

```diff
@@ -0,0 +1,163 @@
+Neither will enrichment tables.
+
+## Pain
+
+The vast majority of Vector's components rely on an external system to be 
+setup and available. This greatly increases the amount of setup required to
+demo or develop any part of Vector that relies on these systems.
+
+## Proposal
+
+### User Experience
+
+In order to enable Demo mode, simply add the configuration to the Vector
+component:
+
+```toml
+demo.enabled = true
+demo.num_records = 10
+demo.frequency_secs = 1
+```
+
+If the component is a source this will result in the source generating a random
+record of data (once per `frequency`) that represents typical data that the 
+source would generate.
+
+`num_records` - the maximum number of records to generate. If not specified
+                records are generated until Vector is killed.
+`frequency_secs` - the gap between generated records. If not specified Vector
+                   generates records as quickly as possible.
+                
+If neither are specified it defaults to generating records once per second 
+ad infinitum.
+
+These options will also be available globally which sets demo mode for all
+components. To keep things simple, global options override any options set on
+individual components (in future, should it be possible to specify the shape of
+data produced these settings will need to be at the component level).
+
+Any configuration must be valid for a component in demo mode even when the
+configuration is not used.
+
+### Implementation
+
+#### Sources
+
+A new method can be added to the `SourceConfig` trait:
+
+```
+fn generate_demo_data(&self) -> Event
+```
+
+Implementers use this to generate a random Event.
+
+A library of data generating functions will be provided that will provide
+random data from a fixed list of data.
+
+#### Sinks
+
+Sinks in demo mode do nothing. No data will be logged - `vector tap` can be
+used to inspect output sent to the sink if required. 
+
+Sinks still need to emit metrics as if they were live.
+
+All data received by the sink will be acknowledged as successfully delivered.
+
+Healthchecks are skipped.
+
+The config settings for `log_namespace` should be respected in demo mode.
+
+#### Reloading 
+
+Special care must be taken to ensure that reloading a config will move a 
+component in and out of demo mode as required
+
+#### Disk buffers
+
+Disk buffers must be disabled when in demo mode to avoid potentially 
+polluting live data with demo data.
+
+## Rationale
+
+Turning demo mode on for a component (source or sink) allows that component 
+to run without accessing any external services. This is useful for a number
+of scenarios:
+
+- whilst configuring Vector it will be much more convenient to do so without 
+  relying on setting up the external services as well.
+- when demonstrating Vector to potential users, it isn't always possible to 
+  set up the external systems needed.
+- observability pipelines needs a way to generate sample data to indicate to
```

Comment:
I'm not convinced this is under the remit of demo mode.

Demo mode is focused on generating data. Observing it is a separate task. I am coming around to the idea that expanding `vector tap` or similar to enable this would be the way forward here. That would provide further benefits beyond OP.